### ISSUE TITLE
implementation buggy for scalar_input

### ISSUE THREAD

**Issue by mostafarohani** [ibab/tensorflow-wavenet#182]
I've been working with this implementation of wavenet using scalar inputs(a.ka. not one hot encoding), since I need an invertible front end that allows me to backprop to the input. I have noticed a couple issues so far: 

1) `predict_proba_incremental` does not have an implementation for scalar_inputs, so I think there should be a flag in generate.py that checks that if `scalar_inputs` is set to true, it will default to `predict_proba`

2.) `predict_proba` is buggy for scalar inputs, it doesn't add the "batch dimension" and therefore causal conv layers will complain. 

3.) The loss function doesn't make sense for scalar inputs. Currently it is doing: 

```
shifted = tf.slice(encoded, [0, 1, 0],
                                   [-1, tf.shape(encoded)[1] - 1, -1])
shifted = tf.pad(shifted, [[0, 0], [0, 1], [0, 0]])
shifted = tf.reshape(shifted, [-1, self.quantization_channels])
prediction = tf.reshape(raw_output,
                                        [-1, self.quantization_channels])
loss = tf.nn.softmax_cross_entropy_with_logits(
                        prediction,
                        shifted)
reduced_loss = tf.reduce_mean(loss)
```

this only makes sense for one hot encoding. First off, a `quantization_channels` size greater than 1 does not make sense for scalar inputs, and we should probably not allow that configuration and complain to the user. Secondly, this loss function is not appropriate since a multinomial loss makes sense when you have scores for different classes, in our case different quantized values. I suggest we add a simple `l2_loss` for the scalar input case for now.

**Comment by jyegerlehner** [comment]
Yes there are bugs around scalar input in the current master. I haven't carefully considered each of your points so I might still be missing something, but I think [this branch](https://github.com/jyegerlehner/tensorflow-wavenet/commits/fix-scalar-input) fixes scalar input. 

I'm not sure the part of the loss function you cite is a bug: even with `scalar_input=true`, the output is a one-hot out of a softmax, so the input still needs to be mu-law-encoded and made into a one-hot so it can serve as the target in the cross-entropy loss.

**Comment by fehiepsi** [comment]
Right now, the fast generation with `scalar_input` is true will raise error. For second point, you can work on @jyegerlehner 's above branch (or this [branch](https://github.com/jyegerlehner/tensorflow-wavenet/tree/fullcorpus_lg), I am working on this branch to fix bug), they fixed problems with slow generation. For the third point, I agree with jyegerlehner that it is not a bug. We quantize input to calculate loss for prediction. I also have thought about `l2_loss`, but then throwing the idea away when I reread WaveNet's paper, where they mentioned that they use log-likelihood at the beginning of Section 2.
